src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,416 @@
+        }
+
+        /// <returns>Positive length or negative value for invalid name</returns>
+        internal static int GetFullTypeNameLength(ReadOnlySpan<char> input, out bool isNestedType)
+        {
+            isNestedType = false;
+
+            // NET 6+ guarantees that MemoryExtensions.IndexOfAny has worst-case complexity
+            // O(m * i) if a match is found, or O(m * n) if a match is not found, where:
+            //   i := index of match position
+            //   m := number of needles
+            //   n := length of search space (haystack)
+            //
+            // Downlevel versions of .NET do not make this guarantee, instead having a
+            // worst-case complexity of O(m * n) even if a match occurs at the beginning of
+            // the search space. Since we're running this in a loop over untrusted user
+            // input, that makes the total loop complexity potentially O(m * n^2), where
+            // 'n' is adversary-controlled. To avoid DoS issues here, we'll loop manually.
+
+#if NET8_0_OR_GREATER
+            int offset = input.IndexOfAny(_endOfFullTypeNameDelimitersSearchValues);
+            if (offset < 0)
+            {
+                return input.Length; // no type name end chars were found, the whole input is the type name
+            }
+
+            if (input[offset] == EscapeCharacter) // this is very rare (IL Emit or pure IL)
+            {
+                offset = GetUnescapedOffset(input, startOffset: offset); // this is slower, but very rare so acceptable
+            }
+#else
+            int offset = GetUnescapedOffset(input, startOffset: 0);
+#endif
+            isNestedType = offset > 0 && offset < input.Length && input[offset] == '+';
+            return offset;
+
+            static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
+            {
+                int offset = startOffset;
+                for (; offset < input.Length; offset++)
+                {
+                    char c = input[offset];
+                    if (c == EscapeCharacter)
+                    {
+                        offset++; // skip the escaped char
+
+                        if (offset == input.Length || // invalid name that ends with escape character
+                            !NeedsEscaping(input[offset])) // invalid name, escapes a char that does not need escaping
+                        {
+                            return -1;
+                        }
+                    }
+                    else if (NeedsEscaping(c))
+                    {
+                        break;
+                    }
+                }
+                return offset;
+            }
+
+            static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
+        }
+
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+            int offset = fullName.LastIndexOfAny('.', '+');
```

Comment:
Recall this comment elsewhere in the file:

https://github.com/dotnet/runtime/blob/d63365fb7293b497a0bc7f17592c3d6439f1d451/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs#L97-L107

Is `GetName` called in a loop? If so, the loop could be an `O(n ^ 2)` operation due to the call to `LastIndexOfAny`.

Reply:
Great question! It's not, as after the most recent perf changes the names for non-simple type names are lazily-allocated

![image](https://github.com/dotnet/runtime/assets/6011991/b1574a4f-4a2c-40f6-9904-e43aa74f8cde)

So this method is never called during the parsing, but may be called when somebody invokes `TypeName.Name`